### PR TITLE
[vscode] Prompt User for Name When Creating New AIConfig

### PR DESCRIPTION
**Pull request by rholinshead**
[vscode] Prompt User for Name When Creating New AIConfig

# [vscode] Prompt User for Name When Creating New AIConfig

VS Code api for creating new/untitled files does not allow specifying the extension to use for the file on save. This results in untitled files being prompted to save with either .json or .yaml extension, NOT .aiconfig.json or .aiconfig.yaml. This is a big problem, since using the non-aiconfig extension will result in the file not opening with our custom editor next time it's opened.

As a workaround, we can default the untilted filename to untitle.aiconfig.json/yaml and prompt the user for the proper name when creating it (before showing the contents). See #1334 for the alternative, of saving the file immediately as untitled and requiring the user to rename manually if they want afterwards.

https://github.com/lastmile-ai/aiconfig/assets/5060851/57a0978a-e377-4c88-bc9f-48dcfc03c8bb

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/lastmile-ai/aiconfig/pull/1337).
* #1342
* __->__ #1337

### CODE REVIEW

File: cookbooks/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "vscode-aiconfig.pythonInterpreter": "/opt/homebrew/Caskroom/miniconda/base/envs/aiconfig/bin/python"
+}
```
Comment (rossdanlm) [inline cookbooks/.vscode/settings.json:3]: Let's not land these for now since they'll cause merge conflicts

File: vscode-extension/src/extension.ts [hunk 2]
```diff
@@ -233,22 +223,64 @@ async function createNewAIConfig(
   const fileContentBuffer = await vscode.workspace.fs.readFile(fileContentPath);
   const initialContent = fileContentBuffer.toString();
 
-  const doc = await vscode.workspace.openTextDocument({
-    content: initialContent,
-    language: mode,
+  const workspaceUri = vscode.workspace.workspaceFolders
+    ? vscode.workspace.workspaceFolders[0].uri
```
Comment (rossdanlm) [inline vscode-extension/src/extension.ts:227]: Since we're just using `workspaceUri.path` later, let's just use `uri.path` here directly. If workspace folder is defined `.uri.fsPath` must exist and can't be undefined: 

Screenshot 2024-02-24 at 17 26 26

File: vscode-extension/src/extension.ts [hunk 3]
```diff
@@ -233,22 +223,64 @@ async function createNewAIConfig(
...
+  const newConfigName = await vscode.window.showInputBox({
+    prompt: "Enter a name for the new AIConfig file",
+    value: firstAvailableUntitledName,
+    validateInput: (input) => {
+      const nameValidationError = validateConfigName(input, mode);
+      if (nameValidationError) {
+        return nameValidationError;
+      }
+
+      return fs.existsSync(input) ? "File already exists" : null;
```
Comment (rossdanlm) [inline vscode-extension/src/extension.ts:255]: Can we just put this inside of the `validateConfigName()` method to keep it cleaner?

File: vscode-extension/src/extension.ts [hunk 4]
```diff
@@ -233,22 +223,64 @@ async function createNewAIConfig(
...
 
-  //const doc = await vscode.workspace.openTextDocument(untitledUri);
-  await vscode.window.showTextDocument(doc, {
-    preview: false,
-    viewColumn: vscode.ViewColumn.One,
+  const newConfigFilePath = workspaceUri
+    ? path.join(workspaceUri.path, newConfigName)
+    : newConfigName;
+  const newConfigUri = vscode.Uri.file(newConfigFilePath).with({
+    scheme: "untitled",
```
Comment (rossdanlm) [inline vscode-extension/src/extension.ts:263]: Why is this needed? Tried looking it up on [VSCode's github](https://github.com/microsoft/vscode-uri) but not super clear what untitled schema does:

Screenshot 2024-02-24 at 17 31 42

File: vscode-extension/src/util.ts [hunk 5]
```diff
@@ -381,6 +381,26 @@ export async function setupEnvironmentVariables(
   }
 }
 
+export function validateConfigName(name: string, mode: "json" | "yaml") {
+  if (name === "") {
+    return "Name is required";
```
Comment (rossdanlm) [inline vscode-extension/src/util.ts:386]: Nit: "Filename is required"

File: vscode-extension/src/util.ts [hunk 6]
```diff
@@ -381,6 +381,26 @@ export async function setupEnvironmentVariables(
...
+  }
+  if (name.includes(" ")) {
+    return "Name cannot contain spaces";
```
Comment (rossdanlm) [inline vscode-extension/src/util.ts:389]: Why is this an issue?

File: vscode-extension/src/util.ts [hunk 7]
```diff
@@ -381,6 +381,26 @@ export async function setupEnvironmentVariables(
...
+  }
+  if (mode === "json" && !name.endsWith("aiconfig.json")) {
```
Comment (rossdanlm) [inline vscode-extension/src/util.ts:391]: nit: ".aiconfig.json" otherwise things like "myaiconfig.json" are valid

File: vscode-extension/src/util.ts [hunk 8]
```diff
@@ -381,6 +381,26 @@ export async function setupEnvironmentVariables(
...
+    return "Name must end with aiconfig.json";
+  }
+  if (
+    mode === "yaml" &&
+    !name.endsWith("aiconfig.yaml") &&
+    !name.endsWith("aiconfig.yml")
```
Comment (rossdanlm) [inline vscode-extension/src/util.ts:397]: Same about ".aiconfig.y[aml|ml]"

**Review by rossdanlm — APPROVED** [review]
lgtm but please address comments before landing

File: cookbooks/.vscode/settings.json (hunk 1 above)
Comment (rholinshead) [reply cookbooks/.vscode/settings.json:3]: Oh, whoops - thanks. Didn't see that sneak in

File: vscode-extension/src/extension.ts (hunk 4 above)
Comment (rholinshead) [reply vscode-extension/src/extension.ts:263]: From https://code.visualstudio.com/api/references/document-selector#:~:text=The%20untitled%20scheme%20is%20for,yet%20been%20saved%20to%20disk.

> The untitled scheme is for new files that have not yet been saved to disk.

Which is technically the case until they save it for the first time

File: vscode-extension/src/util.ts (hunk 6 above)
Comment (rholinshead) [reply vscode-extension/src/util.ts:389]: I'll test it out, but just assumed having spaces in the file name would be problematic as we've seen in other places

Comment (rholinshead) [reply vscode-extension/src/util.ts]: Tested it out, made a config with spaces and it ran and saved fine

**Comment by rholinshead** [comment]
Going to land this so we have something better than current UX. Can fix forward with #1334 if that's the desired ux
